File: ModularMonolithVsMicroservice/common/src/pulse_stats.cpp
#include "pulse_stats.h"

#include <algorithm>

namespace pulsecore {

PulseStatsAccumulator::PulseStatsAccumulator(double sample_rate_hz)
    : sample_rate_hz_(sample_rate_hz) {}

void PulseStatsAccumulator::Add(const pulse::PulseEventBatch& batch) {
  // Columnar events (see pulse.proto): entry k across the parallel
  // arrays is one pulse; iteration walks contiguous doubles.
  const int n = batch.start_sample_size();
  for (int k = 0; k < n; ++k) {
    const double peak = batch.peak_amplitude(k);
    ++count_;
    peak_sum_ += peak;
    duration_sum_ += batch.duration_seconds(k);
    peak_min_ = std::min(peak_min_, peak);
    peak_max_ = std::max(peak_max_, peak);

    if (have_prev_start_) {
      const double gap_samples =
          static_cast<double>(batch.start_sample(k) - prev_start_sample_);
      pri_sum_ += gap_samples / sample_rate_hz_;
      ++pri_count_;
    }
    prev_start_sample_ = batch.start_sample(k);
    have_prev_start_ = true;
  }
}

pulse::PulseSummary PulseStatsAccumulator::Finalize() const {
  pulse::PulseSummary summary;
  summary.set_pulse_count(count_);
  if (count_ > 0) {
    summary.set_mean_peak_amplitude(peak_sum_ / static_cast<double>(count_));
    summary.set_mean_duration_seconds(duration_sum_ / static_cast<double>(count_));
    summary.set_min_peak_amplitude(peak_min_);
    summary.set_max_peak_amplitude(peak_max_);
  }
  if (pri_count_ > 0) {
    summary.set_mean_pri_seconds(pri_sum_ / static_cast<double>(pri_count_));
  }
  return summary;
}

}  // namespace pulsecore

File: ModularMonolithVsMicroservice/python/pulsecore/pulse_stats.py
"""Running summary-statistics accumulator -- Python port of
common/src/pulse_stats.cpp / common/include/pulse_stats.h.
"""
from pulsecore import pulse_pb2


class PulseStatsAccumulator:
    def __init__(self, sample_rate_hz: float):
        self._sample_rate_hz = sample_rate_hz
        self._count = 0
        self._peak_sum = 0.0
        self._duration_sum = 0.0
        self._peak_min = float("inf")
        self._peak_max = float("-inf")
        self._pri_sum = 0.0
        self._pri_count = 0
        self._have_prev_start = False
        self._prev_start_sample = 0

    def add(self, batch: "pulse_pb2.PulseEventBatch") -> None:
        # Same reasoning as PulseDetector.process(): hoist the running
        # state to locals for the loop, write back once at the end.
        count = self._count
        peak_sum = self._peak_sum
        duration_sum = self._duration_sum
        peak_min = self._peak_min
        peak_max = self._peak_max
        pri_sum = self._pri_sum
        pri_count = self._pri_count
        have_prev_start = self._have_prev_start
        prev_start_sample = self._prev_start_sample
        sample_rate_hz = self._sample_rate_hz

        # Columnar events (see pulse.proto): parallel arrays, bulk-copied
        # out of protobuf once, then plain-list iteration.
        starts = list(batch.start_sample)
        peaks = list(batch.peak_amplitude)
        durations = list(batch.duration_seconds)
        for k in range(len(starts)):
            count += 1
            peak_amplitude = peaks[k]
            peak_sum += peak_amplitude
            duration_sum += durations[k]
            if peak_amplitude < peak_min:
                peak_min = peak_amplitude
            if peak_amplitude > peak_max:
                peak_max = peak_amplitude

            start_sample = starts[k]
            if have_prev_start:
                pri_sum += (start_sample - prev_start_sample) / sample_rate_hz
                pri_count += 1
            prev_start_sample = start_sample
            have_prev_start = True

        self._count = count
        self._peak_sum = peak_sum
        self._duration_sum = duration_sum
        self._peak_min = peak_min
        self._peak_max = peak_max
        self._pri_sum = pri_sum
        self._pri_count = pri_count
        self._have_prev_start = have_prev_start
        self._prev_start_sample = prev_start_sample

    def finalize(self) -> "pulse_pb2.PulseSummary":
        summary = pulse_pb2.PulseSummary()
        summary.pulse_count = self._count
        if self._count > 0:
            summary.mean_peak_amplitude = self._peak_sum / self._count
            summary.mean_duration_seconds = self._duration_sum / self._count
            summary.min_peak_amplitude = self._peak_min
            summary.max_peak_amplitude = self._peak_max
        if self._pri_count > 0:
            summary.mean_pri_seconds = self._pri_sum / self._pri_count
        return summary
